packages/hardhat-core/src/internal/solidity/compiler/downloader.ts

```diff
@@ -24,301 +19,337 @@ export enum CompilerPlatform {
+   *
+   * Returns undefined if the compiler has been downloaded but can't be run.
+   *
+   * This function access the filesystem, but doesn't modify it.
+   */
+  getCompiler(version: string): Promise<Compiler | undefined>;
 }
 
-type CompilerDownloaderOptions = Partial<{
-  download: (url: string, destinationFile: string) => Promise<void>;
-  forceSolcJs: boolean;
-}>;
+/**
+ * Default implementation of ICompilerDownloader.
+ *
+ * Important things to note:
+ *   1. If a compiler version is not found, this downloader may fail.
+ *    1.1. It only re-downloads the list of compilers once every X time.
+ *      1.1.1 If a user tries to download a new compiler after X time has passed
+ *      since its release, they may need to clean the cache, as indicated in the
+ *      error messages.
+ */
+export class CompilerDownloader implements ICompilerDownloader {
+  public static getCompilerPlatform(): CompilerPlatform {
+    switch (os.platform()) {
+      case "win32":
+        return CompilerPlatform.WINDOWS;
+      case "linux":
+        return CompilerPlatform.LINUX;
+      case "darwin":
+        return CompilerPlatform.MACOS;
+      default:
+        return CompilerPlatform.WASM;
+    }
+  }
 
-export class CompilerDownloader {
-  private readonly _download: (
-    url: string,
-    destinationFile: string
-  ) => Promise<void>;
-  private readonly _forceSolcJs: boolean;
+  private static _downloaderPerPlatform: Map<string, CompilerDownloader> =
+    new Map();
 
-  constructor(
-    private readonly _compilersDir: string,
-    options: CompilerDownloaderOptions = {}
+  public static getConcurrencySafeDownloader(
+    platform: CompilerPlatform,
+    compilersDir: string
   ) {
-    this._download = options.download ?? downloadFile;
-    this._forceSolcJs = options.forceSolcJs ?? false;
-  }
+    const key = platform + compilersDir;
 
-  public async isCompilerDownloaded(version: string): Promise<boolean> {
-    const compilerBuild = await this.getCompilerBuild(version);
-    const downloadedFilePath = this._getDownloadedFilePath(compilerBuild);
+    if (!this._downloaderPerPlatform.has(key)) {
+      this._downloaderPerPlatform.set(
+        key,
+        new CompilerDownloader(platform, compilersDir)
+      );
+    }
 
-    return this._fileExists(downloadedFilePath);
+    return this._downloaderPerPlatform.get(key)!;
   }
 
-  public async verifyCompiler(
-    compilerBuild: CompilerBuild,
-    downloadedFilePath: string
-  ) {
-    const ethereumjsUtil = await import("ethereumjs-util");
-
-    const expectedKeccak256 = compilerBuild.keccak256;
-    const compiler = await fsExtra.readFile(downloadedFilePath);
+  private readonly _mutex = new Mutex();
 
-    const compilerKeccak256 = ethereumjsUtil.bufferToHex(
-      ethereumjsUtil.keccak(compiler)
-    );
+  /**
+   * Use
+   * @param _platform
+   * @param _compilersDir
+   * @param _compilerListCachePeriodMs
+   * @param _downloadFunction
+   */
```

Comment:
This looks like you started writing the description of the function but didn't get past the first word :stuck_out_tongue: